### DISCUSSION TITLE
gatsby-source-graphql: Incremental Build + Previews doesn't seem to work

### DISCUSSION THREAD

**Discussion by lucashfreitas** [gatsbyjs/gatsby#32341] [Help]
Hello,

Due this issue https://github.com/gatsbyjs/gatsby/issues/31385 and since `contentful` has a graphql api we have decided on use `gatsby-source-graphql` instead of `gatsby-source-contentful`. We could not rely in our building breaking several times a day as contentful source plugin infer types instead of fetching them by structure. 

I am trying to upgrade to `gatsby 3.`x and have also deployed the website in gatsby cloud but the incremental build/preview feature doesn't seem to work. Is that a limitation from `gatsby-source-graphql` as it seems to by pass gatsby data cache or something else?

Thanks,

**Comment by lucashfreitas** [comment]
up

**Comment by LekoArts** [comment]
`gatsby-source-graphql` doesn't support those features.

`gatsby-source-graphql` skips the Gatsby Node model altogether, and allows you to directly pull data from your GraphQL API. At first this seems really attractive as things just work, but as your site grows beyond a handful of pages, a few problems arise.

- There's no way to use Gatsby Preview effectively as the Preview instance will have to run every query in your Gatsby site again. This means in many cases you would need to wait nearly the length of an entire uncached build for your preview to show up.
- There's no way to cache data. This means incremental builds can't work their magic, but it also means regular builds are very slow. Every content change in your API requires all the data to be refetched, making things very slow.

I'm afraid you'll need to wait for https://github.com/gatsbyjs/gatsby/issues/31385 or manually create the schema types to override the inference.

**Reply by lucashfreitas** [reply]
Thanks @LekoArts . Should we add this limitation in the plugin docs as for website content creators preview and incremental build is a critical feature, so people would know that choosing the plugin they won't be able to use those features. 

Thanks again.
